File: 2-add-two-numbers/add-two-numbers.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next
class Solution:
    def addTwoNumbers(self, l1: Optional[ListNode], l2: Optional[ListNode]) -> Optional[ListNode]:
        dummy = ListNode(0)
        cur = dummy
        carry = 0
        while l1 != None or l2 != None or carry != 0:
            l1val = l1.val if l1 else 0
            l2val = l2.val if l2 else 0
            total = l1val + l2val + carry
            carry = total // 10
            newNode = ListNode(total % 10)
            cur.next = newNode
            cur = cur.next
            l1 = l1.next if l1 else None
            l2 = l2.next if l2 else None
        return dummy.next

File: 2-add-two-numbers/add-two-numbers.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode addTwoNumbers(ListNode l1, ListNode l2) {
        int digit = 0;
        ListNode p1 = l1, p2 = l2, dummy = new ListNode(), p3 = dummy;
        while (p1 != null || p2 != null) {
            int v1 = 0, v2 = 0;
            if (p1 != null) {
                v1 = p1.val;
                p1 = p1.next;
            }
            if (p2 != null) {
                v2 = p2.val;
                p2 = p2.next;
            }
            int value = (v1 + v2 + digit) % 10;
            digit = (v1 + v2 + digit) / 10;
            p3.next = new ListNode(value);
            p3 = p3.next;
        }
        if (digit != 0) {
            p3.next = new ListNode(digit);
        }
        return dummy.next;
    }
}